Metrics: noise stacktrace if Prometheus closes the connection due to a timeout

**Describe the bug**
if Prometheus closes the connection due to a timeout or some other reason you will see these errors in the logs and this is very noisy.
We can simply report a better message in the logs.

10:35:18.642 [prometheus-stats-58-1] ERROR org.apache.pulsar.broker.stats.prometheus.PrometheusMetricsServlet - Failed to generate prometheus stats
org.eclipse.jetty.io.EofException: null
	at org.eclipse.jetty.io.ChannelEndPoint.flush(ChannelEndPoint.java:279) ~[jetty-io-9.4.43.v20210629.jar:9.4.43.v20210629]
	at org.eclipse.jetty.io.WriteFlusher.flush(WriteFlusher.java:422) ~[jetty-io-9.4.43.v20210629.jar:9.4.43.v20210629]
	at org.eclipse.jetty.io.WriteFlusher.write(WriteFlusher.java:277) ~[jetty-io-9.4.43.v20210629.jar:9.4.43.v20210629]
	at org.eclipse.jetty.io.AbstractEndPoint.write(AbstractEndPoint.java:381) ~[jetty-io-9.4.43.v20210629.jar:9.4.43.v20210629]
	at org.eclipse.jetty.server.HttpConnection$SendCallback.process(HttpConnection.java:826) ~[jetty-server-9.4.43.v20210629.jar:9.4.43.v20210629]
	at org.eclipse.jetty.util.IteratingCallback.processing(IteratingCallback.java:241) ~[jetty-util-9.4.43.v20210629.jar:9.4.43.v20210629]
	at org.eclipse.jetty.util.IteratingCallback.iterate(IteratingCallback.java:223) ~[jetty-util-9.4.43.v20210629.jar:9.4.43.v20210629]
	at org.eclipse.jetty.server.HttpConnection.send(HttpConnection.java:550) ~[jetty-server-9.4.43.v20210629.jar:9.4.43.v20210629]
	at org.eclipse.jetty.server.HttpChannel.sendResponse(HttpChannel.java:915) ~[jetty-server-9.4.43.v20210629.jar:9.4.43.v20210629]
	at org.eclipse.jetty.server.HttpChannel.write(HttpChannel.java:987) ~[jetty-server-9.4.43.v20210629.jar:9.4.43.v20210629]
	at org.eclipse.jetty.server.HttpOutput.channelWrite(HttpOutput.java:285) ~[jetty-server-9.4.43.v20210629.jar:9.4.43.v20210629]
	at org.eclipse.jetty.server.HttpOutput.channelWrite(HttpOutput.java:269) ~[jetty-server-9.4.43.v20210629.jar:9.4.43.v20210629]
	at org.eclipse.jetty.server.HttpOutput.write(HttpOutput.java:897) ~[jetty-server-9.4.43.v20210629.jar:9.4.43.v20210629]
	at org.apache.pulsar.broker.stats.prometheus.PrometheusMetricsGenerator.generate(PrometheusMetricsGenerator.java:127) .... 4.1.60.Final.jar:4.1.60.Final]
	at java.lang.Thread.run(Thread.java:829) [?:?]
Caused by: java.io.IOException: Broken pipe
	at sun.nio.ch.FileDispatcherImpl.writev0(Native Method) ~[?:?]
	at sun.nio.ch.SocketDispatcher.writev(SocketDispatcher.java:51) ~[?:?]
	at sun.nio.ch.IOUtil.write(IOUtil.java:182) ~[?:?]
	at sun.nio.ch.IOUtil.write(IOUtil.java:130) ~[?:?]
	at sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:493) ~[?:?]
	at java.nio.channels.SocketChannel.write(SocketChannel.java:507) ~[?:?]
	at org.eclipse.jetty.io.ChannelEndPoint.flush(ChannelEndPoint.java:273) ~[jetty-io-9.4.43.v20210629.jar:9.4.43.v20210629]
	... 23 more

**Expected behavior**
The logs should not be full of this spammy log

**Additional context**
The error may occour if you have many functions in the broker and collecting the metrics is too slow.

--- a/pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/PrometheusMetricsServlet.java
+++ b/pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/PrometheusMetricsServlet.java
@@ -85,14 +85,24 @@ public class PrometheusMetricsServlet extends HttpServlet {
                 if (e instanceof EOFException) {
                     // NO STACKTRACE
                     log.error("Failed to send metrics, "
-                            + "likely the client closed the connection due to a timeout ({} ms elapsed): {}",
+                            + "likely the client or this server closed the connection due to a timeout ({} ms elapsed): {}",
                              time, e + "");
                 } else {
                     log.error("Failed to generate prometheus stats, {} ms elapsed", time, e);
                 }
                 res.setStatus(HttpStatus.INTERNAL_SERVER_ERROR_500);
             } finally {
-                context.complete();
+                long end = System.currentTimeMillis();
+                long time = end - start;
+                try {
+                    context.complete();
+                } catch (IllegalStateException e) {
+                    // this happens when metricsServletTimeoutMs expires
+                    // java.lang.IllegalStateException: AsyncContext completed and/or Request lifecycle recycled
+                    log.error("Failed to generate prometheus stats, "
+                                    + "this is likely due to metricsServletTimeoutMs: {} ms elapsed",
+                            time, e + "");
+                }
             }
         }));
     }

--- a/pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/PrometheusMetricsServlet.java
+++ b/pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/PrometheusMetricsServlet.java
@@ -85,8 +85,8 @@ public class PrometheusMetricsServlet extends HttpServlet {
                 if (e instanceof EOFException) {
                     // NO STACKTRACE
                     log.error("Failed to send metrics, "
-                            + "likely the client or this server closed the connection due to a timeout ({} ms elapsed): {}",
-                             time, e + "");
+                            + "likely the client or this server closed "
+                            + "the connection due to a timeout ({} ms elapsed): {}", time, e + "");
                 } else {
                     log.error("Failed to generate prometheus stats, {} ms elapsed", time, e);
                 }

--- a/pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/PrometheusMetricsServlet.java
+++ b/pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/PrometheusMetricsServlet.java
@@ -20,6 +20,7 @@ package org.apache.pulsar.broker.stats.prometheus;
 
 import static org.apache.bookkeeper.mledger.util.SafeRun.safeRun;
 import io.netty.util.concurrent.DefaultThreadFactory;
+import java.io.EOFException;
 import java.io.IOException;
 import java.util.LinkedList;
 import java.util.List;
@@ -70,6 +71,7 @@ public class PrometheusMetricsServlet extends HttpServlet {
         AsyncContext context = request.startAsync();
         context.setTimeout(metricsServletTimeoutMs);
         executor.execute(safeRun(() -> {
+            long start = System.currentTimeMillis();
             HttpServletResponse res = (HttpServletResponse) context.getResponse();
             try {
                 res.setStatus(HttpStatus.OK_200);
@@ -77,11 +79,19 @@ public class PrometheusMetricsServlet extends HttpServlet {
                 PrometheusMetricsGenerator.generate(pulsar, shouldExportTopicMetrics, shouldExportConsumerMetrics,
                         shouldExportProducerMetrics, splitTopicAndPartitionLabel, res.getOutputStream(),
                         metricsProviders);
-                context.complete();
-
             } catch (Exception e) {
-                log.error("Failed to generate prometheus stats", e);
+                long end = System.currentTimeMillis();
+                long time = end - start;
+                if (e instanceof EOFException) {
+                    // NO STACKTRACE
+                    log.error("Failed to send metrics, " +
+                            "likely the client closed the connection due to a timeout ({} ms elapsed): {}",
+                             time, e + "");
+                } else {
+                    log.error("Failed to generate prometheus stats, {} ms elapsed", time, e);
+                }
                 res.setStatus(HttpStatus.INTERNAL_SERVER_ERROR_500);
+            } finally {
                 context.complete();
             }
         }));

--- a/pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/PrometheusMetricsServlet.java
+++ b/pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/PrometheusMetricsServlet.java
@@ -84,8 +84,8 @@ public class PrometheusMetricsServlet extends HttpServlet {
                 long time = end - start;
                 if (e instanceof EOFException) {
                     // NO STACKTRACE
-                    log.error("Failed to send metrics, " +
-                            "likely the client closed the connection due to a timeout ({} ms elapsed): {}",
+                    log.error("Failed to send metrics, "
+                            + "likely the client closed the connection due to a timeout ({} ms elapsed): {}",
                              time, e + "");
                 } else {
                     log.error("Failed to generate prometheus stats, {} ms elapsed", time, e);

